File: Capitulo3_VRCardboard - JAVA/app/src/main/java/com/upv/pm_2022/sep_dic/capitulo3_vrcardboard/Eye.java
package com.upv.pm_2022.sep_dic.capitulo3_vrcardboard;

public class Eye {
    private final float[] mEyeView;
    private final Viewport mViewport;
    private final FieldOfView mFov;
    private volatile boolean mProjectionChanged;
    private float[] mPerspective;
    private float mLastZNear;
    private float mLastZFar;
    
    public Eye(final int type) {
        super();
        this.mEyeView = new float[16];
        this.mViewport = new Viewport();
        this.mFov = new FieldOfView();
        this.mProjectionChanged = true;
    }
    
    public float[] getEyeView() {
        return this.mEyeView;
    }
    
    public float[] getPerspective(final float zNear, final float zFar) {
        if (!this.mProjectionChanged && this.mLastZNear == zNear && this.mLastZFar == zFar) {
            return this.mPerspective;
        }
        if (this.mPerspective == null) {
            this.mPerspective = new float[16];
        }
        this.getFov().toPerspectiveMatrix(zNear, zFar, this.mPerspective, 0);
        this.mLastZNear = zNear;
        this.mLastZFar = zFar;
        this.mProjectionChanged = false;
        return this.mPerspective;
    }
    
    public Viewport getViewport() {
        return this.mViewport;
    }
    
    public FieldOfView getFov() {
        return this.mFov;
    }
    
    public void setProjectionChanged() {
        this.mProjectionChanged = true;
    }
}

File: VRCardboardKotlin/app/src/main/java/com/vrcardboardkotlin/Eye.kt
package com.vrcardboardkotlin

import kotlin.concurrent.Volatile

class Eye(type: Int) {
    val eyeView: FloatArray = FloatArray(16)
    val viewport: Viewport = Viewport()
    val fov: FieldOfView = FieldOfView()

    @Volatile
    private var mProjectionChanged: Boolean = true
    private var mPerspective: FloatArray = FloatArray(16)
    private var mLastZNear = 0f
    private var mLastZFar = 0f

    fun getPerspective(zNear: Float, zFar: Float): FloatArray {
        if (!mProjectionChanged && mLastZNear == zNear && mLastZFar == zFar) {
            return mPerspective
        }
        fov.toPerspectiveMatrix(zNear, zFar, mPerspective, 0)
        mLastZNear = zNear
        mLastZFar = zFar
        mProjectionChanged = false
        return mPerspective
    }

    fun setProjectionChanged() {
        mProjectionChanged = true
    }
}

